packages/plugins/scripts-runner/src/index.ts

```diff
@@ -0,0 +1,178 @@
+
+  name = 'ScriptExecutionError';
+
+  constructor(public target: string, public innerError: Error) {
+    super();
+    this.message = `Script '${path.basename(target)}' failed to execute. Original error: ${innerError}`;
+  }
+}
+
+export default class ScriptsRunnerPlugin {
+
+  embark: Embark;
+  private _web3: Web3 | null = null;
+
+  constructor(embark) {
+    this.embark = embark;
+    // TODO: it'd be wonderful if Embark called `registerCommandHandlers()` for us
+    this.registerCommandHandlers();
+  }
+
+  registerCommandHandlers() {
+    this.embark.events.setCommandHandler(ScriptsRunnerCommand.Execute, this.execute.bind(this));
+  }
+
+  private get web3() {
+    return (async () => {
+      if (!this._web3) {
+        const provider = await this.embark.events.request2('blockchain:client:provider', 'ethereum');
+        this._web3 = new Web3(provider);
+      }
+      return this._web3;
+    })();
+  }
+
+  private async execute(target: string, options: ExecuteOptions, callback: Callback<any>) {
+    const targetPath = !path.isAbsolute(target) ? dappPath(target) : target;
+    try {
+      const fstat = await fsPromises.stat(targetPath);
+      if (fstat.isDirectory()) {
+        const dependencies = await this.getScriptDependencies();
+        await this.executeAll(targetPath, dependencies, callback);
+      } else if (fstat.isFile()) {
+        const dependencies = await this.getScriptDependencies();
+        await this.executeSingle(targetPath, dependencies, callback);
+      } else {
+        callback(new UnsupportedTargetError(fstat));
+      }
+    } catch (e) {
+      callback(e);
+    }
+  }
+
+  private async executeSingle(target: string, dependencies: ScriptDependencies, callback: Callback<any>) {
+    const scriptToRun = require(target);
+    this.embark.logger.info(`.. ${path.basename(target)} running....`);
+    try {
+      const result = await scriptToRun(dependencies);
+      this.embark.logger.info(`.. finished.`);
+      callback(null, result);
+    } catch (e) {
+      const error = e instanceof Error ? e : new Error(e);
+      callback(new ScriptExecutionError(target, error));
+    }
+  }
+
+  private async executeAll(target: string, dependencies: ScriptDependencies, callback: Callback<any>) {
+    const files = await fsPromises.readdir(target);
+    const results: any[] = [];
+
+    eachSeries(files, (file, cb) => {
+      const targetPath = !path.isAbsolute(target) ? dappPath(target, file) : path.join(target, file);
+
+      stat(targetPath, (err, fstat) => {
+        if (err) {
+          return cb(err);
+        }
+        if (fstat.isFile()) {
+          return this.executeSingle(targetPath, dependencies, (e, result) => {
+            if (e) {
+              return cb(e);
+            }
+            results.push(result);
+            cb();
+          });
+        }
+        cb();
+      });
+    }, (err) => {
+      if (err) {
+        return callback(err);
+      }
+      callback(null, results);
+    });
```

Comment:
Wasn't sure if there was a `eachSeries` that aggregates all "results" from each iteratee callback. Maybe @jrainville knows an API that does that, then we can get rid off the custom `results` array.

Reply:
See my question re: promise-and-callback in another comment.

But regardless, another option is to use the JS/TS built-in `async function*` inside `async executeAll`, iterate over the async iterator with `for await () {}` and then return when the iteration is finished.

Here's a rough sketch of revisions to `executeAll` to use `async function*`. It assumes `executeSingle` will return a promise and that `execute` will be responsible for passing the return value of `executeAll` or `executeSingle` via the callback to `execute` rather than having `executeAll/Single` fire the callback.

```ts
import { readdir, stat } from 'fs/promises';

private async executeAll(options: ExecuteOptions) {
  const target = options.target;
  const files = await readdir(target);
  const results: any[] = [];

  async function* scripts() {
    for (const file of files) {
      const targetPath = !path.isAbsolute(target) ? dappPath(target, file) : path.join(target, file);
      const fstat = await stat(targetPath);
      if (!fstat.isFile()) continue;
      yield {
        target: targetPath,
        dependencies: options.dependencies,
        forceTracking: options.forceTracking
      };
    }
  }

  for await (const script of scripts) {
    results.push(await this.executeSingle(script));
  }

  return results;
}
```

Reply:
After lunch I thought of a slightly better way to do it (see below). Main difference is that stats are collected and filtered up-front and in parallel, and if something blows up during that process then no scripts are run (which is probably the desirable behavior). Also, it's not strictly necessary to use `async function*`, i.e. `function*` will work fine in this case:

```ts
import { readdir, stat } from 'fs/promises';

private async executeAll(options: ExecuteOptions) {
  const target = options.target;
  const files = await readdir(target);
  const results: any[] = [];

  const scripts = (await Promise.all(
    files.map(file => {
      const targetPath = !path.isAbsolute(target) ? dappPath(target, file) : path.join(target, file);
      return { targetPath, stats: stat(targetPath) };
    })
  ))
    .filter(({stats}) => stats.isFile())
    .map(({targetPath}) => ({
      target: targetPath,
      dependencies: options.dependencies,
      forceTracking: options.forceTracking
    }));

  function* execSerial() {
    for (const script of scripts) {
      yield this.executeSingle(script);
    }
  }

  for await (const result of execSerial()) {
    results.push(result);
  }

  return results;
}
```

Reply:
Related:

[Iterator Helpers](https://github.com/tc39/proposal-iterator-helpers) is a Stage 2 TC39 proposal and a polyfill is [provided by core-js](https://github.com/zloirock/core-js/#iterator-helpers). We already depend on a version of `core-js` that has those polyfills, but babel doesn't apply those particular ones automatically since they're not shipped proposals. We could load them manually.

The point of Iterator Helpers is to make working with Iterators and AsyncIterators as convenient as working with arrays. In my opinion Iterator Helpers is basically the big missing piece that prevents wider usage of generators/iterators (sync or async) since without them you can't practically reach for `.map()`, `.filter()`, `.reduce()`. But once Iterator Helpers are a JS/TS language built-in, I think generators/iterators will see a surge in popular usage.

In the second of my two code examples above, instead of using a results array and a `for await` loop at the end, you could do:

```js
import 'core-js/proposals/iterator-helpers';
// ...
return execSerial().toArray();
// ^ .toArray() returns a promise
```

The scripts would execute in order, one-by-one, and the resolved return value would be an array of results.

One of the inspirations for Iterator Helpers proposal was/is IxJS. You can do something equivalent with it to avoid using a results array:

```js
import { from as ixFrom, toArray as ixToArray } from 'ix/asynciterable';
// ...
return ixToArray(ixFrom(execSerial());
// ^ ixToArray() returns a promise
```

As with Iterator Helpers, the scripts would execute in order, one-by-one, and the resolved return value would be an array of results.

Reply:
I appreciate your suggestions. I do feel though that both look far more complex (not saying they are), to the eye. I'd aim for making this particular part of the code a bit more digestible but especially the second suggestion just looks much harder to parse for me.

Any more opinions from @embarklabs/core ?

Happy to go down that route, otherwise.

Reply:
@PascalPrecht sorry for not answering earlier about your question to aggregate results.
There is `concat` or `concatSeries` that you can use: http://caolan.github.io/async/v3/docs.html#concat

Reply:
@PascalPrecht fair enough.

Even sticking with it as is, I would be grateful to you and other @embarklabs/core members if we can give some thought to the problem more generally.

We're all busy and I don't want to take up time unnecessarily, but I think it's something we should think about.

As a team we've discussed, on numerous occasions, migrating away from the async library. At some point in the future, even if not right now, that means:

1. No longer writing any new code that makes use of the async library (unless it's new/refactored code situated in the middle of existing usage and it's not a good time to refactor existing).
2. Over time refactoring existing code to use language built-ins and/or libraries that provide sugar around built-ins (or Observables, which aren't built-ins yet, and sugar for them).

---

Point **(2)** needs a caveat insofar as callbacks are certainly a built-in language feature and the async library is arguably sugar for callbacks. However, callback wrangling, even with the help of the async library, tends to be messier and it's easy to misunderstand what's going on. Case in point:

**me**:
> > Related to that, will the callback fire multiple times in the context of executeAll?

**Pascal**:
> It won't. `executeAll()` passes an `eachSeries()` callback to `executeSingle`. The callback passed to `executeAll` is only called once when the series is done...

I simply misread the code, and on another look I might have realized I read it wrong, but with multiple layers of inner and outer callbacks, and mixing with `asyn function` for the convenience of `await`, it's easy to make a mistake reading and/or writing such code. Mixing with `async` also invites the anti-pattern of an API both firing a callback and returning a promise (though in private APIs the dangers can be somewhat mitigated).

Regarding point **(1)**, if we really are going to stop using the async library for new code and over time refactor existing code, then the longer we wait on **(1)** the more refactoring we'll have to do in the future.

---

Generally, as a though experiment, let's assume async library is off the table...

Given a `Collection` to process (let's assume it's an array) and a set of `Steps`...

For each `step` in `Steps`:
* When each `item` in `Collection` can be run through `step` concurrently, ask:
  * Does `step` do something asynchronous?
    * If so, can use `Promise.all` or `Promise.allSettled` together with `.map()` on `Collection` (possibly with the help of `require('util').promisify` in `step`).
    * If not, can simply use `.map()`, `.reduce()`, `filter()`, `.forEach()`, or lodash methods on `Collection`, depending on what needs to be done; because `step` doesn't do something asynchronous concurrent processing isn't possible anyway.
* When each `item` in `Collection` should be run through `step` serially, ask:
  * Does `step` do something asynchronous?
    * If so, setup a generator that yields promises for each `item` in `Collection` and then iterate with `for await` or use async iteration sugar.
    * If not, can simply use `.map()`, `.reduce()`, etc. on `Collection`, depending on what needs to be done, because `step` doesn't do something asynchronous.

Often times, if you have multiple successive asynchronous steps that require serial processing, they can be performed within in a single generator. Conveniently, both `await` and `yield` are available in async generators (`async function*`).

---

Less generally, @PascalPrecht could you give some feedback on which parts of my second example was hard to parse?  Was it the first part with the longish `await Promise.all().filter().map()`? Or was it the second part that began with `function* execSerial`?

I ask because *if* we do begin a migration away from the async library, it would be helpful to know what aspects of using promise helpers and generators/iterators are confusing.

Also, I'll note that the first part could be written a bit differently to make it clearer. And if we use async interation sugar (whether via core-js or something else), the `.filter` and final `.map` steps of the first part could be combined in a very clean way with the second part, while still getting the performance boost of doing all the `stat` calls concurrently in a `Promise.all`.

Thanks for time spent reading this and thinking about it. My only goal here is for the Embark codebase to become more robust and easier to maintain over time.

Reply:
I learned later today that it's possible to manually load `core-js` polyfills in a way that doesn't introduce globals or modify prototype chains. The way to do it is to to use the `core-js-pure` package, which could be added as a dep (not devDep) in `packages/plugins/scripts-runner/package.json`.

Revised example, hopefully nicer than before (we don't even have to manually create a generator :tada:):

```ts
import AsyncIterator from 'core-js-pure/features/async-iterator';
import * as fs from 'fs';

const { readdir, stat } = fs.promises;

// ...

private async executeAll(options: ExecuteOptions) {
  const target = options.target;
  const files = await readdir(target);

  const scripts = await Promise.all(
    files.map(file => {
      const targetPath = !path.isAbsolute(target) ? dappPath(target, file) : path.join(target, file);
      return { target: targetPath, stats: stat(targetPath) };
    })
  );

  return AsyncIterator.from(scripts)
    .filter(({stats}) => stats.isFile())
    .map(({target}) => ({
      target,
      dependencies: options.dependencies,
      forceTracking: options.forceTracking
    }))
    .map(script => this.executeSingle(script))
    .toArray();
}
```

There are [several other methods](https://github.com/zloirock/core-js/#iterator-helpers) available on instances of `AsyncIterator` that may be useful.

And here's some JS you can throw in a Node.js REPL for experimentation; it follows the same workflow as above. The point of `doubleCheck`, `tripleCheck` below is to show that the async iteration chain processes serially, i.e. members of the collection being processed never jump ahead of the others.

```
$ mkdir -p ~/temp/tryme && cd ~/temp/tryme
$ npm init -y && npm i core-js-pure && node
```

```js
const AsyncIterator = require('core-js-pure/features/async-iterator');
const { readdir, stat } = require('fs').promises;

const timer = (ms) => new Promise(resolve => { setTimeout(resolve, ms); });

const timedReturn = async (n) => { await timer(n); return n; };

const times = [300, 22, 1, 888, 999, 100, 2, 0];

const doubleCheck = [];
const tripleCheck = [];

const finished = (async () => {

  const promisedTimes = await Promise.all(times.map(t => Promise.resolve(t)));

  return AsyncIterator.from(promisedTimes)
    .filter(n => !(n % 2))
    .map(n => { doubleCheck.push(n); return n; })
    .map(n => timedReturn(n))
    .map(n => { tripleCheck.push(n); return n; })
    .toArray();
})();

(async () => {
  console.log(await finished);
  console.log(doubleCheck);
  console.log(tripleCheck);
})();
```

Reply:
> Less generally, @PascalPrecht could you give some feedback on which parts of my second example were hard to parse? Was it the first part with the longish await Promise.all().filter().map()? Or was it the second part that began with function* execSerial?

I must admit that it probably boils down to mostly unfamiliarity with the new APIs described in your comments. This is often the case when new concepts and ideas are introduced. I imagine the same would happen to many team members when we would introduce observables in our APIs. The upside to them are often not the problem. It's rather that it comes at the cost of introducing a different layer instead that needs to be learned.

> Regarding point (1), if we really are going to stop using the async library for new code and over time refactor existing code, then the longer we wait on (1) the more refactoring we'll have to do in the future.

I fully agree and I'll give your suggestion a shot. Thanks for being resistant.